ui/lib/core/addon/components/confirm.js

```diff
@@ -0,0 +1,70 @@
+import Component from '@ember/component';
+import { computed } from '@ember/object';
+import { htmlSafe } from '@ember/template';
+import layout from '../templates/components/confirm';
+import { next } from '@ember/runloop';
+
+/**
+ * @module Confirm
+ * `Confirm` components prevent users from performing actions they do not intend to by showing a confirmation message as an overlay. This is a contextual component that should always be rendered with a `Message` which triggers the message.
+ *
+ * @example
+ * ```js
+ * <div class="box">
+ * <Confirm as |c|>
+ *   <c.Message
+ *     @id={{item.id}}
+ *     @triggerText="Delete"
+ *     @message="This will permanently delete this secret and all its vesions."
+ *     @onConfirm={{action "delete" item "secret"}}
+ *     />
+ * </Confirm>
+ * </div>
+ * ```
+ */
+
+export default Component.extend({
+  layout,
+  openTrigger: null,
+  height: 0,
+  focusTrigger: null,
+  style: computed('height', function() {
+    return htmlSafe(`height: ${this.height}px`);
+  }),
+  wormholeReference: null,
+  wormholeId: computed(function() {
+    return `confirm-${this.elementId}`;
+  }),
+  didInsertElement() {
+    this.set('wormholeReference', this.element.querySelector(`#${this.wormholeId}`));
+  },
+  didRender() {
+    this.updateHeight();
+  },
+  updateHeight: function() {
+    let height;
+    height = this.openTrigger
+      ? this.element.querySelector('.confirm-overlay').clientHeight
+      : this.element.querySelector('.confirm').clientHeight;
+    this.set('height', height);
+  },
+  actions: {
+    onTrigger: function(itemId, e) {
+      this.set('openTrigger', itemId);
+
+      // store a reference to the trigger so we can focus the element
+      // after clicking cancel
+      this.set('focusTrigger', e.target);
+      this.updateHeight();
+    },
+    onCancel: function() {
+      this.set('openTrigger', '');
+      this.updateHeight();
+
+      next(() => {
+        this.focusTrigger.focus();
+        this.set('focusTrigger', null);
```

Comment:
we may want to do this in the `willDestroy` hook too

Reply:
i can add that in, but out of curiosity, how does adding
```
next(() => {
        this.focusTrigger.focus();
        this.set('focusTrigger', null);
```
to the `willDestroy` help here? i initially thought it'd help prevent memory leaks, but it doesn't appear to be a problem. i noticed while using the ember inspector that the `Confirm`'s `focusTrigger` does correctly get reset to `null` after cancelling the confirmation message, and re-triggering that same trigger does result in no `focusTrigger` being set yet. is there another scenario i should test?

Reply:
Yeah, I was thinking memory leaks as well since you can click "disable" and then click outside of the popup without hitting cancel. Thinking about it again though, the button is also unrendered when the popup goes away so there wouldn't be a dangling reference so you don't need to null out the reference.